### ISSUE TITLE
Performance decimalize

### ISSUE THREAD

**Issue by ubmarco** [jsvine/pdfplumber#197]
Hi,

TLDR: Can pdfplumber shift to using floats instead of decimals for better performance?

I'm writing a Python 3.6/3.7/3.8 library on top of pdfplumber that uses the page cropping feature a lot.
When bigger PDFs are analyzed I realized that my lib becomes really slow.
I did some research using the Python `cProfile` module for the pdfplumber 2-page test file `tests/pdfs/issue-33-lorem-ipsum.pdf`.
The profiler call is `python -m cProfile -s time examples/lib_usage_example.py`, so it is sorting for internal time (the total time spent in the given function, excluding time made in calls to sub-functions).
```
         17572011 function calls (17495455 primitive calls) in 5.741 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
5795639/5790657    1.818    0.000    2.573    0.000 utils.py:112(decimalize)
  1439296    1.275    0.000    3.781    0.000 utils.py:276(point_inside_bbox)
  6039739    0.746    0.000    0.756    0.000 {built-in method builtins.isinstance}
  1799120    0.327    0.000    4.108    0.000 utils.py:288(<genexpr>)
   359824    0.261    0.000    4.605    0.000 utils.py:281(obj_inside_bbox_score)
   360232    0.237    0.000    4.344    0.000 {built-in method builtins.sum}
   360492    0.128    0.000    4.732    0.000 utils.py:333(<genexpr>)
      109    0.070    0.001    0.070    0.001 {method 'read' of '_io.BufferedReader' objects}
      668    0.043    0.000    4.781    0.007 utils.py:330(n_points_intersecting_bbox)
```
The biggest obstacle in performance is the `decimalize` function. It consumes 32% of overall program runtime. This scales up to a lot of time when running on big PDFs.

I can think of 3 reasons that may lead to this:
1. floats are natively hardware supported (FPU) while decimal types are emulated (using C extensions for Python 3). [This](https://stackoverflow.com/questions/41453307/decimal-python-vs-float-runtime) SO answer says decimals are ~2.5 times slower than floats for Python 3. 
2. The conversion code itself just has to run, without decimal there is no need to convert. pdfminer uses float, so not converting would be the best option I see. I know some numeric advantages that comes with initially using decimals, but the source is float and the conversion changes the numbers, so I don't see the advantages of decimals in this use case.
3. I haven't checked that yet, but maybe there's room for improvement in the conversion logic.

What's your opinion on that? If we can agree on a common proposal, I'm also willing to spend further analysis time and/or send a PR because it's a major impediment for me.

**Comment by jsvine** [comment]
Hi @ubmarco, and thank you for raising this issue. I agree: It'd be great to find ways to improve PDFPlumber's performance. And if there's a way to do it safely by reducing conversions, that'd be great. 

Some background: PDFPlumber uses `Decimal` because of the problems that floats both cause regarding precision. (`Decimal` is also more conceptually faithful to the PDF specification, but that's a secondary issue.) 

There are many cases in PDF extraction when being able to match positions precisely is important — for instance, when selecting all objects *fully within* a bounding box. The imprecision of floats (especially when the result of adding two floating point numbers) can mean the difference between an object falling within, or outside a bounding box. This, essentially, is the problem we encounter:

```
>>> 0.1 + 0.2
0.30000000000000004
```

If all PDFs used the same degree of precision — e.g., to `0.01` — then we could probably use floats and handle the imprecision with rounding. But, unfortunately, not all PDFs do. So using `Decimal` seemed like the most accurate, consistent way of handling PDF object positioning.

However, it might not be the __only__ solution. If you, or others, have proposals for maintaining precision while improving performance, I'm entirely open to it.

Thanks again,
Jeremy

**Comment by ubmarco** [comment]
Hi @jsvine and thanks for the fast reply.
I understand the reasons to use decimals (coming from an embedded background). I also read through the Python docs for Decimal.
I just wonder, how this problem can disappear when interfacing with a library that provides floats.

If, for example, the user defines a crop page and pdfplumber wants to determine which characters are part of the bbox, then it will look at pdfminer char positions defined as float. To run the conversion, pdfplumber can either convert pdfminer's float to decimal or the user crop bbox to float.
That shouldn't make a difference.
To enable the comparison, I would go with an epsilon/delta range anyway.

Any thoughts?

**Comment by ubmarco** [comment]
Btw, there is also the [math.isclose](https://docs.python.org/3/library/math.html#math.isclose) function that was added in Python 3.5. And float precision in Python is double, so I guess it's enough to represent what a PDF needs. I don't see general issues solving the float comparison and representation problems. And we have to live somehow with pdfminer that returns floats.

Anyway, I did an analysis with the decimalize function that eats most of the runtime.
When looking at above profiling numbers, the `ncalls` of `decimalize` are `5795639/5790657`.
That means that from overall 5795639 calls only 4982 are recursive calls. That means almost all calls are forwarded to _decimalize. So the only reason why decimalize could be so expensive that it does an `isinstance` check - which is known to have a poor performance.

To measure the amount of types that are coming in, I changed the code to:
```
conversion_types = {}
def decimalize(v, q = None):
    global conversion_types
    if type(v) not in conversion_types:
        conversion_types[type(v)] = 1
    else:
        conversion_types[type(v)] += 1
    # If tuple/list passed, bulk-convert
    if isinstance(v, (tuple, list)):
        return type(v)(decimalize(x, q) for x in v)
    else:
        return _decimalize(v, q)
```
The result for `conversion_types`:
```
{
    <class 'int'>: 193,
    <class 'list'>: 273,
    <class 'tuple'>: 1379,
    <class 'decimal.Decimal'>: 5760793,
    <class 'float'>: 33001,
}
```
That's astonishing, 99.4% of all converted numbers are already Decimal.
So I changed the function again an introduced a shortcut for decimals (that is already contained in `_decimalize`):
```
def decimalize(v, q = None):
    # If already a decimal, just return itself
    if isinstance(v, Decimal):
        return v
    # If tuple/list passed, bulk-convert
    if isinstance(v, (tuple, list)):
        return type(v)(decimalize(x, q) for x in v)
    else:
        return _decimalize(v, q)
```
The result is awesome:
```
         17606841 function calls (17530283 primitive calls) in 4.308 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  1439296    1.276    0.000    2.433    0.000 utils.py:280(point_inside_bbox)
5795639/5790657    0.867    0.000    1.228    0.000 utils.py:113(decimalize)
  6074361    0.353    0.000    0.363    0.000 {built-in method builtins.isinstance}
  1799120    0.310    0.000    2.742    0.000 utils.py:292(<genexpr>)
```
So decimalize goes down from 1.818s internal time to 0.867, which is a reduction by almost 50%.
The call count is still the same, that makes sense.

Shall I send a PR for this?

**Comment by ubmarco** [comment]
Hi,
I was wrong on [my comment](https://github.com/jsvine/pdfplumber/issues/197#issuecomment-609868564).

> If, for example, the user defines a crop page and pdfplumber wants to determine which characters are part of the bbox, then it will look at pdfminer char positions defined as float. To run the conversion, pdfplumber can either convert pdfminer's float to decimal or the user crop bbox to float.
> That shouldn't make a difference.

I just checked and float conversion to Decimal is actually lossless:
```
>>> import decimal
>>> from decimal import Decimal
>>> dvar = Decimal(fvar)
>>> dvar
Decimal('0.1000000000000000055511151231257827021181583404541015625')
>>> from decimal import Decimal
>>> fvar = float(0.1)
>>> print('{0:0.70f}'.format(fvar))
0.1000000000000000055511151231257827021181583404541015625000000000000000
>>> dvar = Decimal(fvar)
>>> dvar
Decimal('0.1000000000000000055511151231257827021181583404541015625')
```
So I think the current implementation is fine, maybe the call count to the `decimalize` function could be reduced by keeping track of which numbers where already converted and implementing above proposal.

**Comment by jsvine** [comment]
A big thanks, @ubmarco, for taking such a close look at the bottleneck here. Your solution of checking `if isinstance(v, Decimal):` early looks like a great one. Yes, please do submit a PR!

Also, I don't think that we'll ever need to deal with subclasses of `Decimal`, so I think it's safe there to replace `isinstance(v, Decimal)` with `type(v) == Decimal` if it improves performance.

And to continue this part of the conversation:

> If, for example, the user defines a crop page and pdfplumber wants to determine which characters are part of the bbox, then it will look at pdfminer char positions defined as float. To run the conversion, pdfplumber can either convert pdfminer's float to decimal or the user crop bbox to float.

Part of the problem is that `.crop` is not the only place where two positions might be compared. We need to compare positions for `extract_words`, and other functions. I'm wary of an implementation that would require every function like this to potentially check/convert types.

**Comment by jsvine** [comment]
I've added the performance improvement you suggested, @ubmarco, and credited you in the commit and the README acknowledgments. Thanks again for suggesting it! Now available in v0.5.19.

**Comment by ubmarco** [comment]
Thanks a lot for the credits. And sorry for being unresponsive regarding the PR, the times are a bit busy here... I'll close the issue now.
